packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/WorkflowController.java

```diff
@@ -327,6 +328,51 @@ public ResponseEntity<Workflow> selectOutput(
 		return updated.map(ResponseEntity::ok).orElseGet(() -> ResponseEntity.notFound().build());
 	}
 
+	@PostMapping("/{id}/update-position")
+	@Secured(Roles.USER)
+	@Operation(summary = "Update node and edge positions")
+	@ApiResponses(
+		value = {
+			@ApiResponse(
+				responseCode = "200",
+				description = "Workflow updated.",
+				content = @Content(
+					mediaType = MediaType.APPLICATION_JSON_VALUE,
+					schema = @io.swagger.v3.oas.annotations.media.Schema(implementation = Workflow.class)
+				)
+			),
+			@ApiResponse(responseCode = "404", description = "Workflow could not be found", content = @Content),
+			@ApiResponse(responseCode = "500", description = "There was an issue updating the workflow", content = @Content)
+		}
+	)
+	public ResponseEntity<Workflow> updatePositions(
+		@PathVariable("id") final UUID id,
+		@RequestBody final WorkflowPositions payload,
+		@RequestParam(name = "project-id", required = false) final UUID projectId
+	) {
+		final Schema.Permission permission = projectService.checkPermissionCanRead(
+			currentUserService.get().getId(),
+			projectId
+		);
+
+		final Optional<Workflow> workflow = workflowService.getAsset(id, permission);
+		final Optional<Workflow> updated;
+
+		try {
+			workflowService.updatePositions(workflow.get(), payload);
+			updated = workflowService.updateAsset(workflow.get(), projectId, permission);
```

Comment:
I do not see a `workflow.isEmpty()` test here before accessing the Optional Workflow.

Reply:
That is a good point - we haven't been doing this kind of safety checks in WorkflowController. 

I will do a separate cleanup pass for the class to keep this PR more closely contained with the original intent.